File: samples/triggers/ado-comment-watcher.py
#!/usr/bin/env python3
"""
ado-comment-watcher.py

Same trigger as ado-comment-watcher.ts, written in Python.

Demonstrates that the Conductor trigger protocol is language-agnostic:
  - stdin:  one JSON envelope
  - env:    CONDUCTOR_MCP_URL, CONDUCTOR_MCP_SECRET (required for Mode B POSTs),
            ADO_ORG, ADO_BEARER_TOKEN (preferred) or ADO_PAT (fallback)
  - stdout: { state, systemMessage } JSON  (no `callback` field — Mode B)
  - exit:   0 ok, 2 blocking error, other non-blocking error

Two execution modes are supported by the protocol — this script uses Mode B.

  Mode A (NOT used here): respond via stdout.
    Set a single `callback: { body: {...} }` object on the JSON response.
    Conductor delivers that one entry to env["callback_url"]. Mode A's
    `callback` is SINGULAR (at most one delivery per run).

  Mode B (USED HERE): script POSTs to env["callback_url"] directly during
    the run, once per detected event. Required when a single run may need
    to deliver more than one event. The stdout response carries only
    `{ state, systemMessage }` — no `callback`.

To use, register this trigger in .conductor/triggers.json with:
  "command": "python3 $CONDUCTOR_PROJECT_DIR/.conductor/triggers/ado-comment-watcher.py"

Standard library only — no `requests`, no `httpx`, no SDK. Just `urllib`,
`json`, `os`, `sys`, `base64`. Runs on any Python 3.8+.
"""

from __future__ import annotations

import base64
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from typing import Any
from datetime import datetime


# ============================================================================
# Config from env
# ============================================================================

ADO_ORG = os.environ.get("ADO_ORG", "")
ADO_PAT = os.environ.get("ADO_PAT", "")
ADO_BEARER_TOKEN = os.environ.get("ADO_BEARER_TOKEN", "")

# Mode B requires this for the Authorization header on direct callback POSTs.
CONDUCTOR_MCP_SECRET = os.environ.get("CONDUCTOR_MCP_SECRET", "")


def ado_auth_header() -> str:
    """Bearer token preferred; PAT (basic auth) as fallback."""
    if ADO_BEARER_TOKEN:
        return f"Bearer {ADO_BEARER_TOKEN}"
    if ADO_PAT:
        return "Basic " + base64.b64encode(f":{ADO_PAT}".encode()).decode()
    raise RuntimeError("ADO_BEARER_TOKEN or ADO_PAT environment variable required")


# ============================================================================
# I/O helpers
# ============================================================================

def write_stdout(response: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(response))


def blocking_error(reason: str) -> None:
    sys.stderr.write(f"{reason}\n")
    sys.exit(2)


# ============================================================================
# ADO client (stdlib urllib — bearer token preferred, basic-auth with PAT fallback)
# ============================================================================

def list_pr_comments(repo: str, pr_id: int, since_id: int) -> list[dict[str, Any]]:
    """Return ADO PR comments with id > since_id, sorted ascending by id."""
    if not ADO_ORG:
        raise RuntimeError("ADO_ORG environment variable required")

    url = (
        f"https://dev.azure.com/{ADO_ORG}/_apis/git/repositories/"
        f"{urllib.parse.quote(repo)}/pullRequests/{pr_id}/threads"
        f"?api-version=7.1-preview.1"
    )

    req = urllib.request.Request(url, headers={"Authorization": ado_auth_header()})
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            body = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"ADO {e.code}: {e.read().decode(errors='replace')}") from e

    comments: list[dict[str, Any]] = []
    for thread in body.get("value", []):
        for c in thread.get("comments", []):
            if c.get("commentType") == "system":
                continue                       # skip status changes, votes
            cid = c.get("id")
            if cid is None or cid <= since_id:
                continue                       # already-seen
            author = c.get("author") or {}
            comments.append({
                "id": cid,
                "content": c.get("content") or "",
                "author": {
                    "uniqueName": author.get("uniqueName"),
                    "displayName": author.get("displayName"),
                },
                "publishedDate": c.get("publishedDate") or datetime.utcnow().isoformat() + "Z",
            })

    comments.sort(key=lambda c: c["id"])       # monotonic order
    return comments


# ============================================================================
# Prompt construction — the actual "intelligence" of this trigger
# ============================================================================

def comment_to_prompt(pr_id: int, comment: dict[str, Any]) -> str:
    """Translate a raw ADO comment into a human-framed instruction for the agent."""
    author = (
        comment["author"].get("displayName")
        or comment["author"].get("uniqueName")
        or "a reviewer"
    )
    try:
        at = datetime.fromisoformat(comment["publishedDate"].replace("Z", "+00:00")).strftime(
            "%Y-%m-%d %H:%M"
        )
    except Exception:
        at = comment["publishedDate"]

    quoted = "\n> ".join(comment["content"].split("\n"))

    return "\n".join([
        f"New comment on PR {pr_id} from {author} ({at}):",
        "",
        f"> {quoted}",
        "",
        "Look at this comment in the context of your current review.",
        "If it's a question, draft a clear answer grounded in the diff.",
        "If it's a change request, draft a plan and ask the user via approval.request before applying.",
        "If it's affirming, acknowledge briefly and continue.",
        "Consult the respond-to-pr-comment template for tone and structure.",
    ])


# ============================================================================
# Mode B — POST a single callback directly to env["callback_url"].
# ============================================================================

def post_callback(callback_url: str, body: dict[str, Any]) -> None:
    if not CONDUCTOR_MCP_SECRET:
        raise RuntimeError(
            "CONDUCTOR_MCP_SECRET environment variable required for Mode B callback POSTs"
        )
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        callback_url,
        data=data,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {CONDUCTOR_MCP_SECRET}",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            # Drain so the connection is reusable; we don't need the body.
            resp.read()
    except urllib.error.HTTPError as e:
        text = e.read().decode(errors="replace") if hasattr(e, "read") else "<no body>"
        raise RuntimeError(f"callback POST {e.code}: {text}") from e


def body_for_comment(pr_id: int, comment: dict[str, Any]) -> dict[str, Any]:
    return {
        "prompt": comment_to_prompt(pr_id, comment),
        "context": {
            "source": "ado",
            "kind": "pr.commented",
            "pr_id": pr_id,
            "comment_id": comment["id"],
        },
    }


# ============================================================================
# Main
# ============================================================================

def main() -> None:
    stdin_text = sys.stdin.read()
    if not stdin_text.strip():
        write_stdout({"systemMessage": "No stdin envelope received."})
        return

    try:
        env = json.loads(stdin_text)
    except json.JSONDecodeError as e:
        blocking_error(f"Invalid JSON on stdin: {e}")

    # Hydrate state with defaults
    state = {
        "prId": 0,
        "repo": "",
        "lastCommentId": 0,
        "selfUser": "",
        **(env.get("state") or {}),
    }

    if not state["prId"] or not state["repo"]:
        blocking_error("state.prId and state.repo must be set when the trigger is registered")

    callback_url = env.get("callback_url")
    if not callback_url:
        blocking_error("env.callback_url missing — required for Mode B live POSTs")

    # Note: env["trigger_data_dir"] is available as a per-trigger scratch dir at
    # <project_dir>/.conductor/triggers/<trigger_id>/data/. This script doesn't
    # need it (lastCommentId fits in state), but a richer trigger could write
    # blobs there, e.g.:
    #   import pathlib
    #   data = pathlib.Path(env["trigger_data_dir"])
    #   data.mkdir(parents=True, exist_ok=True)
    #   (data / f"comment-{c['id']}.json").write_text(json.dumps(c))

    posted = 0

    payload = env.get("payload") or {}
    resource = payload.get("resource") or {}
    external_comment = resource.get("comment")
    external_pr = resource.get("pullRequest") or {}
    external_pr_matches = external_pr.get("pullRequestId") == state["prId"]

    # ----- Real-time path: ADO service hook delivered the comment -----
    if env.get("fired_by") == "external" and external_comment and external_pr_matches:
        author = external_comment.get("author") or {}
        if (
            external_comment.get("commentType") != "system"
            and author.get("uniqueName") != state["selfUser"]
        ):
            c = {
                "id": external_comment["id"],
                "content": external_comment["content"],
                "author": {"uniqueName": author.get("uniqueName")},
                "publishedDate": external_comment.get("publishedDate")
                or datetime.utcnow().isoformat() + "Z",
            }
            post_callback(callback_url, body_for_comment(state["prId"], c))
            state["lastCommentId"] = c["id"]
            posted += 1

        write_stdout({
            "state": state,
            "systemMessage": (
                f"Forwarded 1 comment from ADO service hook (PR {state['prId']})."
                if posted
                else "Skipped self/system comment."
            ),
        })
        return

    # ----- Cron / manual / agent: poll ADO for new comments -----
    new_comments = list_pr_comments(state["repo"], state["prId"], state["lastCommentId"])

    for c in new_comments:
        if c["author"].get("uniqueName") == state["selfUser"]:
            continue
        post_callback(callback_url, body_for_comment(state["prId"], c))
        state["lastCommentId"] = c["id"]
        posted += 1

    write_stdout({
        "state": state,
        "systemMessage": (
            f"Forwarded {posted} new comment(s) on PR {state['prId']} "
            f"(fired_by={env.get('fired_by')})."
            if posted
            else f"No new comments on PR {state['prId']} (fired_by={env.get('fired_by')})."
        ),
    })


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        raise
    except Exception as err:
        sys.stderr.write(f"{type(err).__name__}: {err}\n")
        sys.exit(1)

File: samples/triggers/ado-comment-watcher.ts
#!/usr/bin/env tsx
/**
 * ado-comment-watcher.ts
 *
 * A working trigger that listens for new comments on Azure DevOps PRs and
 * asks the agent to address each one.
 *
 * REFERENCE IMPLEMENTATION of Mode B (live POSTs during the run).
 *
 * Two execution modes are supported by the protocol — this script uses Mode B.
 *
 *   Mode A (NOT used here): respond via stdout.
 *     Build a single `callback: { body: {...} }` object on the JSON response
 *     and write it on stdout. Clawdevbox delivers that one entry to the
 *     trigger's pre-bound `env.callback_url`. The script makes zero HTTP calls.
 *     Mode A's `callback` is SINGULAR — at most one delivery per run, suitable
 *     for one-shot triggers or final/summary actions.
 *
 *   Mode B (USED HERE): script POSTs to env.callback_url directly during the
 *     run, once per detected event. Required when a single run may need to
 *     deliver more than one event. The stdout response carries only `{ state,
 *     systemMessage }` — no `callback` field, since the live POSTs already
 *     delivered the work.
 *
 *   No-op: just `{ "state": <unchanged> }` and exit 0. No HTTP calls.
 *
 * The script does NOT think about routing. Clawdevbox pre-bound
 * `env.callback_url` to the right action when this trigger was registered
 * (in this case: append + wake the subscriber thread). The URL itself
 * carries the routing context; the script just delivers the prompt.
 *
 * Two firing modes (orthogonal to A/B above), same handler:
 *   - fired_by="external": ADO service hook delivered the comment payload directly
 *   - fired_by="cron|manual|agent": no payload; poll ADO since lastCommentId
 *
 * Zero dependencies beyond Node 18+ built-in fetch.
 */

// ============================================================================
// Types
// ============================================================================

type FiredBy = 'external' | 'cron' | 'manual' | 'agent';

interface TriggerEnvelope {
  trigger_event_name: 'TriggerFired';
  trigger_id: string;
  run_id: string;
  fired_by: FiredBy;
  fired_at: number;
  cwd: string;
  /** Workspace root — the user's project directory. */
  project_dir: string;
  /**
   * Per-trigger scratch directory at
   *   <project_dir>/.clawdevbox/triggers/<trigger_id>/data/
   * The script can write/read any files here to maintain data across runs
   * (caches, downloaded artifacts, anything too large for `state`).
   * Clawdevbox creates it on first run and cleans it up when the trigger is
   * deleted (for hot triggers, when the subscriber thread terminates).
   */
  trigger_data_dir: string;
  subscriber_thread_id: string | null;
  /**
   * Pre-bound callback URL. In Mode B (this script), the script POSTs each
   * detected event directly to this URL during the run.
   */
  callback_url: string;
  state: WatcherState;
  payload: AdoServiceHookPayload | null;
}

interface WatcherState {
  /** PR id this trigger watches (set once at trigger.upsert). */
  prId: number;
  /** ADO repo name. */
  repo: string;
  /** Last comment id we forwarded; advances monotonically. */
  lastCommentId: number;
  /** Skip our own comments to avoid reply loops. */
  selfUser: string;
}

interface AdoServiceHookPayload {
  resource?: {
    pullRequest?: { pullRequestId: number };
    comment?: {
      id: number;
      content: string;
      commentType?: 'system' | 'text';
      author?: { uniqueName?: string; displayName?: string };
      publishedDate?: string;
    };
  };
}

interface AdoComment {
  id: number;
  content: string;
  author: { uniqueName?: string; displayName?: string };
  publishedDate: string;
}

/** Body the script POSTs to env.callback_url for each detected event (Mode B). */
interface CallbackBody {
  prompt: string;
  context?: Record<string, unknown>;
}

interface TriggerResponse {
  state?: WatcherState;
  systemMessage?: string;
  decision?: 'ok' | 'block';
  reason?: string;
}

// ============================================================================
// I/O helpers
// ============================================================================

async function readStdin(): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) chunks.push(chunk as Buffer);
  return Buffer.concat(chunks).toString('utf8');
}

function writeStdout(response: TriggerResponse): void {
  process.stdout.write(JSON.stringify(response));
}

function blockingError(reason: string): never {
  process.stderr.write(`${reason}\n`);
  process.exit(2);
}

// ============================================================================
// ADO client (raw HTTP — bearer token preferred, basic-auth with PAT fallback)
//
// Auth env vars (one of these is required):
//   ADO_BEARER_TOKEN  — AAD access token (e.g. from `az account get-access-token`).
//                       Preferred when both are set.
//   ADO_PAT           — Azure DevOps personal access token (basic auth).
// ============================================================================

const ADO_ORG = process.env.ADO_ORG ?? '';
const ADO_PAT = process.env.ADO_PAT ?? '';
const ADO_BEARER_TOKEN = process.env.ADO_BEARER_TOKEN ?? '';

/** Mode B requires this for the Authorization header on direct callback POSTs. */
const CLAWDEVBOX_MCP_SECRET = process.env.CLAWDEVBOX_MCP_SECRET ?? '';

function adoAuthHeader(): string {
  if (ADO_BEARER_TOKEN) return `Bearer ${ADO_BEARER_TOKEN}`;
  if (ADO_PAT) return `Basic ${Buffer.from(`:${ADO_PAT}`).toString('base64')}`;
  throw new Error('ADO_BEARER_TOKEN or ADO_PAT env var required');
}

async function listPrComments(repo: string, prId: number, sinceId: number): Promise<AdoComment[]> {
  if (!ADO_ORG) throw new Error('ADO_ORG env var required');

  const url =
    `https://dev.azure.com/${ADO_ORG}/_apis/git/repositories/` +
    `${encodeURIComponent(repo)}/pullRequests/${prId}/threads?api-version=7.1-preview.1`;

  const res = await fetch(url, { headers: { Authorization: adoAuthHeader() } });
  if (!res.ok) throw new Error(`ADO ${res.status}: ${await res.text()}`);

  const body = (await res.json()) as {
    value?: Array<{ comments?: Array<Record<string, unknown>> }>;
  };

  const comments: AdoComment[] = [];
  for (const thread of body.value ?? []) {
    for (const c of thread.comments ?? []) {
      if (c.commentType === 'system') continue;
      const id = c.id as number;
      if (id <= sinceId) continue;
      const author = c.author as { uniqueName?: string; displayName?: string } | undefined;
      comments.push({
        id,
        content: (c.content as string) ?? '',
        author: { uniqueName: author?.uniqueName, displayName: author?.displayName },
        publishedDate: (c.publishedDate as string) ?? new Date().toISOString(),
      });
    }
  }

  return comments.sort((a, b) => a.id - b.id);
}

// ============================================================================
// Prompt construction — the actual "intelligence" of this trigger
// ============================================================================

/**
 * Translate a raw ADO comment into a human-framed instruction for the agent.
 * The agent reads this prompt and decides how to respond.
 */
function commentToPrompt(prId: number, comment: AdoComment): string {
  const author = comment.author.displayName ?? comment.author.uniqueName ?? 'a reviewer';
  const at = new Date(comment.publishedDate).toLocaleString();

  return [
    `New comment on PR ${prId} from ${author} (${at}):`,
    ``,
    `> ${comment.content.split('\n').join('\n> ')}`,
    ``,
    `Look at this comment in the context of your current review.`,
    `If it's a question, draft a clear answer grounded in the diff.`,
    `If it's a change request, draft a plan and ask the user via approval.request before applying.`,
    `If it's affirming, acknowledge briefly and continue.`,
    `Consult the respond-to-pr-comment template for tone and structure.`,
  ].join('\n');
}

// ============================================================================
// Mode B — POST a single callback directly to env.callback_url.
// ============================================================================

async function postCallback(callbackUrl: string, body: CallbackBody): Promise<void> {
  if (!CLAWDEVBOX_MCP_SECRET) {
    throw new Error('CLAWDEVBOX_MCP_SECRET env var required for Mode B callback POSTs');
  }
  const res = await fetch(callbackUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${CLAWDEVBOX_MCP_SECRET}`,
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => '<no body>');
    throw new Error(`callback POST ${res.status}: ${text}`);
  }
}

function bodyForComment(prId: number, comment: AdoComment): CallbackBody {
  return {
    prompt: commentToPrompt(prId, comment),
    context: { source: 'ado', kind: 'pr.commented', pr_id: prId, comment_id: comment.id },
  };
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  const stdin = await readStdin();
  if (!stdin.trim()) {
    writeStdout({ systemMessage: 'No stdin envelope received.' });
    return;
  }

  let env: TriggerEnvelope;
  try {
    env = JSON.parse(stdin);
  } catch (err) {
    blockingError(`Invalid JSON on stdin: ${(err as Error).message}`);
  }

  // Hydrate state with defaults (spread first so env.state can override)
  const defaults: WatcherState = { prId: 0, repo: '', lastCommentId: 0, selfUser: '' };
  const state: WatcherState = { ...defaults, ...env.state };

  if (!state.prId || !state.repo) {
    blockingError('state.prId and state.repo must be set when the trigger is registered');
  }

  const callbackUrl = env.callback_url;
  if (!callbackUrl) {
    blockingError('env.callback_url missing — required for Mode B live POSTs');
  }

  // Note: env.trigger_data_dir is available as a per-trigger scratch directory
  // at `<project_dir>/.clawdevbox/triggers/<trigger_id>/data/`. This script
  // doesn't need it (lastCommentId fits in `state`), but a richer trigger
  // could write blobs there, e.g.:
  //   import { writeFile, mkdir } from 'node:fs/promises';
  //   import { join } from 'node:path';
  //   await mkdir(env.trigger_data_dir, { recursive: true });
  //   await writeFile(join(env.trigger_data_dir, `comment-${c.id}.json`), JSON.stringify(c));

  let posted = 0;

  // ----- Real-time path: ADO service hook delivered the comment -----
  const externalComment = env.payload?.resource?.comment;
  const externalPrMatches = env.payload?.resource?.pullRequest?.pullRequestId === state.prId;

  if (env.fired_by === 'external' && externalComment && externalPrMatches) {
    if (
      externalComment.commentType !== 'system' &&
      externalComment.author?.uniqueName !== state.selfUser
    ) {
      const c: AdoComment = {
        id: externalComment.id,
        content: externalComment.content,
        author: { uniqueName: externalComment.author?.uniqueName },
        publishedDate: externalComment.publishedDate ?? new Date().toISOString(),
      };
      await postCallback(callbackUrl, bodyForComment(state.prId, c));
      state.lastCommentId = c.id;
      posted++;
    }

    writeStdout({
      state,
      systemMessage:
        posted > 0
          ? `Forwarded 1 comment from ADO service hook (PR ${state.prId}).`
          : `Skipped self/system comment.`,
    });
    return;
  }

  // ----- Cron / manual / agent: poll ADO for new comments -----
  const newComments = await listPrComments(state.repo, state.prId, state.lastCommentId);

  for (const c of newComments) {
    if (c.author.uniqueName === state.selfUser) continue;
    await postCallback(callbackUrl, bodyForComment(state.prId, c));
    state.lastCommentId = c.id;
    posted++;
  }

  writeStdout({
    state,
    systemMessage:
      posted > 0
        ? `Forwarded ${posted} new comment(s) on PR ${state.prId} (fired_by=${env.fired_by}).`
        : `No new comments on PR ${state.prId} (fired_by=${env.fired_by}).`,
  });
}

main().catch((err) => {
  process.stderr.write((err as Error).stack ?? String(err));
  process.stderr.write('\n');
  process.exit(1);
});
